torch/lib/THD/base/channels/DataChannelTCP.cpp

```diff
@@ -195,49 +195,60 @@ DataChannelTCP::~DataChannelTCP()
 
 
 void DataChannelTCP::listen(std::uint16_t port = 0) {
-  SYSCHECK(_socket = ::socket(PF_INET, SOCK_STREAM, 0))
+  struct addrinfo hints, *res = NULL;
 
-  struct sockaddr_in addr;
-  socklen_t addr_len = sizeof(addr);
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  int err = getaddrinfo(NULL, std::to_string(port).data(), &hints, &res);
+  if (err != 0) {
+    throw std::invalid_argument("cannot find host to listen on");
+  }
 
-  memset(&addr, 0, addr_len);
-  addr.sin_family = AF_INET;
-  addr.sin_port = htons(port);
-  addr.sin_addr.s_addr = INADDR_ANY;
+  // TODO: should we prioritize any `ai_family`?
+  SYSCHECK(_socket = ::socket(res->ai_family, res->ai_socktype, res->ai_protocol))
 
   int optval = 1;
   SYSCHECK(::setsockopt(_socket, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(int)))
-  SYSCHECK(::bind(_socket, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)))
+  SYSCHECK(::bind(_socket, res->ai_addr, res->ai_addrlen))
   SYSCHECK(::listen(_socket, LISTEN_QUEUE_SIZE))
-  SYSCHECK(::getsockname(_socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
 
+  // get listen port
+  struct sockaddr_in addr;
+  socklen_t addr_len = sizeof(addr);
+  SYSCHECK(::getsockname(_socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
   _port = ntohs(addr.sin_port);
 }
 
 
 int DataChannelTCP::connect(const std::string& address, std::uint16_t port,
                             int wait = true) const {
-  struct sockaddr_in addr;
-  socklen_t addr_len = sizeof(addr);
-
-  memset(&addr, 0, addr_len);
-  addr.sin_family = AF_INET;
-  addr.sin_port = htons(port);
-  struct addrinfo *res;
-
-  // get address by host or IP
-  int err = ::getaddrinfo(address.data(), NULL, NULL, &res);
-  if (err == 0) {
-    std::memcpy(
-      &(addr.sin_addr),
-      &(reinterpret_cast<struct sockaddr_in*>(res->ai_addr)->sin_addr),
-      sizeof(struct in_addr)
-    );
-    ::freeaddrinfo(res);
+  struct in6_addr server_addr;
+  struct addrinfo hints, *res = NULL;
+
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_NUMERICSERV;
```

Comment:
This flag says that second parameter of `getaddrinfo` (which is called `service`) should be interpreted as decimal port number (not binary).

"If AI_NUMERICSERV is specified in hints.ai_flags and service is not NULL, then service must point to a string containing a numeric port number"